Refactor event derivation mechanism

NDStrahilevitz opened the pull request:
This PR refactors the `deriveEvent` function and mechanism so that it is:

1. More declarative in syntax
2. Takes less cpu time (reduced by 3.91% of overall time in code path)

# The changes

The `Tracee` object now initializes a map on startup of type `map[int32]map[int32]deriveFn`, this represents a map of a derived EventID to its possible derivations and their corresponding deriving functions.

Then `deriveEvents` can simply check if the current event has any derivations to be done and get the map of all it's derivative functions, instead of calculating that map on each `deriveEvents` call which causes unnecessary GC work due to the map and function variable assignments.

This improves CPU usage in runtime and allows a more declarative approach to defining event derivatives through the map initialization function (`initEventDerivationMap` in `events_derived.go`)

# Performance analysis

In terms of benchmarking I did not see any regression in throughput.
I have generated pprofs of before and after to check if the bottleneck was removed, the graph output shows this:

<img width="1082" alt="image" src="https://user-images.githubusercontent.com/22661609/160284355-82abf924-b884-4db2-9eef-441632adc675.png">

As can be seen we have reduced time in mallocgc through a reduction from `deriveEvents`. As expected we have gained some time back in `mapaccess` however it's easily seen that the difference in time is negative (a gain in cpu time).
I have not seen any new bottleneck arising from this, and I have attached the pprofs so this can be confirmed by others.

pprofs: 
1. [before](https://github.com/aquasecurity/tracee/files/8357725/070.pb.gz)
2. [after](https://github.com/aquasecurity/tracee/files/8357728/refactor.pb.gz)

A diff pprof can be generated as such:
`go tool pprof -http=":4441" -diff_base 070.pb.gz refactor.pb.gz`

mtcherni95 commented on pkg/ebpf/derivations.go:
```diff
@@ -0,0 +1,61 @@
+	"github.com/aquasecurity/tracee/types/trace"
+)
+
+//returns the derived event, wether an event was derived, and a relevant error if failure was due to one
```
typo in `whether`

NDStrahilevitz commented on pkg/ebpf/derivations.go:
```diff
@@ -0,0 +1,61 @@
+//returns the derived event, wether an event was derived, and a relevant error if failure was due to one
+type deriveFn func(trace.Event, EventDefinition) (trace.Event, bool, error)
+
+func deriveContainerCreate(t *Tracee) deriveFn {
```
I see why we would want that but that can easily cause circular dependencies since most derivation function actually depend on it's tracee's object current state, so it's difficult to separate.
I'll see if it's possible since it would be better if we can do it.

NDStrahilevitz replied:
Actually, these aren't exported so there's not much of reason to be worried about usage outside of `ebpf`

mtcherni95 replied:
it is still more correct to have it under internal, you do not what anyways to have them used outside (no one should expose them)

NDStrahilevitz replied:
Resolved offline, this will be done in a later PR to keep it out of this one's scope.

yanivagman replied:
I would wait with putting it under "internal" of the ebpf package.
The ebpf package will eventually handle only ebpf related stuff - bpf maps population, program loading/attachment, bpf filters and perf buffer polling and decoding.
In the future, we can potentially split some of this logic to a new "events" package, and there we can have such "internal" packages

NDStrahilevitz commented on pkg/ebpf/derivations.go:
```diff
@@ -0,0 +1,61 @@
+type deriveFn func(trace.Event, EventDefinition) (trace.Event, bool, error)
+
+func deriveContainerCreate(t *Tracee) deriveFn {
+	return func(event trace.Event, def EventDefinition) (trace.Event, bool, error) {
```
Like a `Derivation` struct? I can't think of a case where we wouldn't immediately want all 3 values so having them returned unpacked is convenient.

yanivagman commented on pkg/ebpf/derivations.go:
```diff
@@ -0,0 +1,72 @@
+package ebpf
```
I don't see a reason to split these functions to a new file.
It is short enough, and will be easier to read if we will keep it in the same file

yanivagman replied:
#1561 and #1583 shouldn't be derived events. I still didn't have the time to review these PRs yet, but from a quick look, the logic there should be moved to process_events.
Even if we will have more derived events in the future, the deriveEvent function is the only generic function there, and all the other deriveFns are the specific ones. By placing the generic deriveEvent function first in the file, I think this should be easier to read and maintain

NDStrahilevitz replied:
Done

yanivagman commented on pkg/ebpf/tracee.go:
```diff
@@ -18,6 +17,8 @@ import (
 	"time"
 	"unsafe"
 
+	"github.com/aquasecurity/tracee/pkg/events/queue"
```
If you move this, let's put it below just before "github.com/aquasecurity/tracee/pkg/events/sorting"

NDStrahilevitz replied:
I think this moved due to automatic formatting, I didn't touch it myself.

NDStrahilevitz replied:
done

yanivagman commented on pkg/ebpf/events_derived.go:
```diff
@@ -6,85 +6,33 @@ import (
-		deriveFns = map[int32]deriveFn{
-			ContainerRemoveEventID: containerRemoveFn,
-		}
+func (t *Tracee) initEventDerivationMap() error {
+	t.eventDerivations = map[int32]map[int32]deriveFn{
+		CgroupMkdirEventID: {
+			ContainerCreateEventID: deriveContainerCreate(t),
+		},
+		CgroupRmdirEventID: {
+			ContainerRemoveEventID: deriveContainerRemoved(t),
+		},
 	}
 
+	return nil
```
If you already made this step to initialize these functions here, you can also save some other runtime steps by adding the derived event "template", which will contain the event id, event name, return value, stack addresses, args, and args num of the event. Then, when deriving the event in runtime, there will be no need to get the event definition and set these fields again

yanivagman replied:
Indeed this is for optimizations purposes.
If you approve that there is no much of performance gain by doing that, then I'm ok with keeping it that way

NDStrahilevitz replied:
There might be a performance gain only if we had an event that is VERY frequently derived (this might be workload dependent as well), and even so a map access is a pretty cheap operation (notice that we access the derivation map for every event and it takes only 0.5% cpu, it is a small map however), so I think this can be considered a premature optimization.

yanivagman reviewed (approved):
Nice refactor, LGTM!

rafaeldtinoco commented:
I'll be taking a look at this and merging (looks good so far). Commits should be squashed and git log should have proper line breaks.. I can do it after review.

rafaeldtinoco commented on pkg/ebpf/events_derived.go:
```diff
@@ -6,87 +6,102 @@ import (
+		},
+		CgroupRmdirEventID: {
+			ContainerRemoveEventID: deriveContainerRemoved(t),
+		},
```
Single liners here are easier to read/understand:

```go
	t.eventDerivations = map[int32]map[int32]deriveFn{
		CgroupMkdirEventID: {ContainerCreateEventID: deriveContainerCreate(t)},
		CgroupRmdirEventID: {ContainerRemoveEventID: deriveContainerRemoved(t)},
	}
```

rafaeldtinoco replied:
I see your point. Kind of makes sense. Thanks.

rafaeldtinoco commented on pkg/ebpf/events_derived.go:
```diff
@@ -6,87 +6,102 @@ import (
-type deriveFn func(EventDefinition) error
+//deriveFn represents a function which takes an event and derives whether another event should result from it
+//
+//returns a derived (or empty) event, whether an event was derived, and an error if relevant
```
You could improve this comment, it is not very clear. Suggestion:

```go
// deriveFn is a function prototype for a function that receives an event as
// argument and produces a new event, depending on the event type received. It
// returns the derived event, a bool indicating if there was indeed a derived
// event created, and an error/nil.
```

NDStrahilevitz replied:
I've rewritten to this:

```
// deriveFn is a function prototype for a function that receives an event as
// argument and may produce a new event if relevant.
// It returns the a derived or empty event, depending on succesful derivation,
// a bool indicating if an event was derived, and an error if one occured.
```

rafaeldtinoco commented on pkg/ebpf/events_derived.go:
```diff
@@ -6,87 +6,102 @@ import (
-			return nil
+/*
+* Derivation functions:
+* Each derivation function initially takes it's runtime dependencies (usually tracee to track its state)
```
Could you clarify what you meant here ?

rafaeldtinoco commented on pkg/ebpf/events_derived.go:
```diff
@@ -6,87 +6,102 @@ import (
+* Derivation functions:
+* Each derivation function initially takes it's runtime dependencies (usually tracee to track its state)
+* It then returns the actual deriveFn which is stored in tracee's derivation map (where it initializes the closure with its own state)
+* From the derivation map we call the relevant scoped deriveFn
```
And we're limited to a single derived function per event type (which is okay as we're currently using one). I wonder if this comment is needed or it would be better to add description to each derive function AND `deriveEvent()` one, with a generic explanation.

rafaeldtinoco reviewed (changes requested):
Change looks very good and, tbh, I think there might be other places where we could benefit of same technique (not to create/re-create function definitions over and over.

I'd like comments to be improved (gave simple suggestion) and each derive function to have its own description (what it is doing and why). This way next derived functions will follow the same guideline.

NDStrahilevitz commented:
> Change looks very good and, tbh, I think there might be other places where we could benefit of same technique (not to create/re-create function definitions over and over.
> 
> I'd like comments to be improved (gave simple suggestion) and each derive function to have its own description (what it is doing and why). This way next derived functions will follow the same guideline.

Thanks for the feedback, I've squashed the commits, added comments and rewritten the others, take a look when you have the time :)

rafaeldtinoco reviewed (approved):
Thanks for the changes. LGTM now.
